Controller is not assignable to parameter of type 'SocketClass'

Hello,

I use your library in my project and everything works but I receive type errors.

controller class:
```typescript
import { Inject, Injectable } from '@decorators/di';
import { Ack, Args, Controller, Event } from '@decorators/socket/src';

import { logger } from '../../../logger';
import { StateService } from '../../../state/service';

@Injectable()
@Controller('/')
export class DoorController {
  constructor(@Inject(StateService) private stateService: StateService) { }

  @Event('api/door')
  getDoor(@Args() message, @Ack() callback) {
    logger.debug(`api/door: ${message.groupAddress}`);
    callback(this.stateService.getDoorOpener(message.groupAddress));
  }
}
```

Error: 
```
src\api\door\controller\index.ts(12,4): error TS2345: Argument of type 'DoorController' is not assignable to parameter of type 'SocketClass'.
  Property '__socket_meta__' is missing in type 'DoorController'.
```

In my index.ts I attach all controllers.

Any ideas how to fix this issue?

BR Jakob

--- a/playground/socket/index.ts
+++ b/playground/socket/index.ts
@@ -5,6 +5,7 @@ import {
   Controller,
   Socket,
   Args,
+  ServerMiddleware,
   Middleware,
   Event,
   IO_MIDDLEWARE
@@ -14,7 +15,7 @@ import { Container, Injectable, Inject, InjectionToken } from '@decorators/di';
 const server = listen(3000);
 const MESSAGE = new InjectionToken('MESSAGE');
 
-class ServerMiddleware implements Middleware {
+class GlobalMiddleware implements ServerMiddleware {
   public use(io, socket, next) {
     console.log('ServerMiddleware');
     next();
@@ -49,7 +50,7 @@ class ConnectionController {
 
 Container.provide([
   { provide: MESSAGE, useValue: 'Socket welcomes user' },
-  { provide: IO_MIDDLEWARE, useClass: ServerMiddleware }
+  { provide: IO_MIDDLEWARE, useClass: GlobalMiddleware }
 ]);
 
 attachControllers(server, [ ConnectionController ]);

--- a/playground/socket/package.json
+++ b/playground/socket/package.json
@@ -1,12 +1,15 @@
 {
   "name": "@decorators/playground-socket",
-  "version": "3.0.0",
+  "version": "3.1.0",
   "description": "node decorators",
-  "main": "index.js",
   "dependencies": {
+    "@decorators/di": ">=1.0.2",
     "@decorators/socket": "../../socket",
-    "@types/socket.io": "1.4.29",
-    "typescript": "2.6.1"
+    "socket.io": ">=2.1.1"
+  },
+  "devDependencies": {
+    "@types/socket.io": ">=1.4.36",
+    "typescript": ">=2.9.2"
   },
   "keywords": [
     "nodejs",
@@ -20,7 +23,7 @@
     "url": "https://github.com/serhiiso/node-decorators/issues"
   },
   "engines": {
-    "node": ">=6.3.0"
+    "node": ">=7.1.0"
   },
   "repository": {
     "type": "git",

--- a/socket/CHANGELOG.md
+++ b/socket/CHANGELOG.md
@@ -1,3 +1,9 @@
+# Socket#3.1.0
+* Changed library structure
+* Fixed interfaces
+* Added new interface to implement for server middleware `ServerMiddleware`, controller `Middleware` interface remains the same
+* Updated versions of dependencies, like `socket.io` and `@decorators/di`
+
 # Socket#3.0.2
 * Fix sorting parameters when applying events
 

--- a/socket/README.md
+++ b/socket/README.md
@@ -52,11 +52,25 @@ class ControllerMiddleware implements Middleware {
   }
 }
 ```
-To register global middleware handler, use **IO_MIDDLEWARE** injection token with **Container** from `@decorators/di` package, like so:
+To register global middleware handler create a class, that implements interface **ServerMiddleware**, like so:
+```typescript
+class GlobalMiddleware implements ServerMiddleware {
+  public use(
+    io: SocketIO.Server | SocketIO.Namespace,
+    socket: SocketIO.Socket,
+    next: Function
+  ) {
+    console.log('GlobalMiddleware');
+    next();
+  }
+}
+```
+
+Then register it by using **IO_MIDDLEWARE** injection token with **Container** from `@decorators/di` package, like so:
 
 ```typescript
 Container.provide([
-  { provide: IO_MIDDLEWARE, useClass: ServerMiddleware }
+  { provide: IO_MIDDLEWARE, useClass: GlobalMiddleware }
 ]);
 ```
 
@@ -64,7 +78,6 @@ Container.provide([
 #### Middleware
 The middleware order :
 * Global Server Middleware (**io.use(...)**)
-* Global Socket middleware (**socket.use(...)**)
 * Controller based middleware (**@Controller(...)**)
 * Event based middleware (**@Event(...)**)
 

--- a/socket/index.ts
+++ b/socket/index.ts
@@ -1,0 +1,1 @@
+export * from './src';

--- a/socket/package.json
+++ b/socket/package.json
@@ -1,15 +1,16 @@
 {
   "name": "@decorators/socket",
-  "version": "3.0.2",
+  "version": "3.1.0",
   "description": "node decorators - decorators for socket.io library",
-  "main": "src/index.js",
+  "main": "lib/index.js",
+  "types": "lib/index.d.ts",
   "dependencies": {
-    "socket.io": ">=2.0.4",
-    "@decorators/di": ">=1.0.0"
+    "socket.io": ">=2.1.1",
+    "@decorators/di": ">=1.0.2"
   },
   "devDependencies": {
-    "@types/socket.io": ">=1.4.31",
-    "typescript": ">=2.6.1"
+    "@types/socket.io": ">=1.4.36",
+    "typescript": ">=2.9.2"
   },
   "keywords": [
     "nodejs",
@@ -23,14 +24,13 @@
     "url": "https://github.com/serhiiso/node-decorators/issues"
   },
   "engines": {
-    "node": ">=6.3.0"
+    "node": ">=7.1.0"
   },
   "repository": {
     "type": "git",
     "url": "https://github.com/serhiiso/node-decorators.git"
   },
   "scripts": {
-    "compile": "tsc",
-    "clean-tsc": "find . -name \"*.js\" -type f -not -path \"./node_modules/*\" -delete && find . -name \"*.js.map\" -type f -not -path \"./node_modules/*\" -delete && find . -name \"*.d.ts\" -type f -not -path \"./node_modules/*\" -not -path \"./types/*\" -delete"
+    "compile": "tsc"
   }
 }

--- a/socket/src/decorators/controller.ts
+++ b/socket/src/decorators/controller.ts
@@ -1,13 +1,13 @@
 import { getMeta, SocketMeta } from '../meta';
-import { Middleware } from '../middleware';
+import { Type } from '../middleware';
 
 /**
  * Defines namespace for the controller and controller-based middleware
  *
  * @param {string} namespace
- * @param {Middleware[]} [middleware]
+ * @param {Type[]} [middleware]
  */
-export function Controller(namespace: string, middleware?: Middleware[]) {
+export function Controller(namespace: string, middleware?: Type[]) {
   return (target): void => {
     const meta: SocketMeta = getMeta(target.prototype);
 

--- a/socket/src/decorators/events.ts
+++ b/socket/src/decorators/events.ts
@@ -1,14 +1,14 @@
 import { getMeta, SocketClass, SocketMeta, EventType } from '../meta';
-import { Middleware } from '../middleware';
+import { Type } from '../middleware';
 
 /**
  * Listener decorator factory, creates listener decorator
  *
  * @param {EventType} type Listener type: io or socket
  * @param {string} event Event name
- * @param {Middleware[]} middleware Event middleware
+ * @param {Type[]} middleware Event middleware
  */
-function makeDecorator(type: EventType, event: string, middleware?: Middleware[]) {
+function makeDecorator(type: EventType, event: string, middleware?: Type[]) {
   return (target: SocketClass, methodName: string, descriptor?: any) => {
     const meta: SocketMeta = getMeta(target);
 
@@ -23,9 +23,9 @@ function makeDecorator(type: EventType, event: string, middleware?: Middleware[]
  *
  * @see **io.on('connection', fn)**
  *
- * @param {Middleware[]} middleware Event middleware
+ * @param {Type[]} middleware Event middleware
  */
-export function Connection(middleware?: Middleware[]) {
+export function Connection(middleware?: Type[]) {
   return makeDecorator(EventType.IO, 'connection', middleware);
 }
 /**
@@ -39,9 +39,9 @@ export const Connect = Connection;
  * @see **io.on**
  *
  * @param {string} event
- * @param {Middleware[]} middleware Event middleware
+ * @param {Type[]} middleware Event middleware
  */
-export function GlobalEvent(event: string, middleware?: Middleware[]) {
+export function GlobalEvent(event: string, middleware?: Type[]) {
   return makeDecorator(EventType.IO, event, middleware);
 }
 
@@ -50,9 +50,9 @@ export function GlobalEvent(event: string, middleware?: Middleware[]) {
  *
  * @see **socket.on('disconnect', fn)**
  *
- * @param {Middleware[]} middleware Event middleware
+ * @param {Type[]} middleware Event middleware
  */
-export function Disconnect(middleware?: Middleware[]) {
+export function Disconnect(middleware?: Type[]) {
   return makeDecorator(EventType.Socket, 'disconnect', middleware);
 }
 
@@ -62,8 +62,8 @@ export function Disconnect(middleware?: Middleware[]) {
  * @see **socket.on**
  *
  * @param {string} event
- * @param {Middleware[]} middleware Event middleware
+ * @param {Type[]} middleware Event middleware
  */
-export function Event(event: string, middleware?: Middleware[]) {
+export function Event(event: string, middleware?: Type[]) {
   return makeDecorator(EventType.Socket, event, middleware);
 };

--- a/socket/src/index.ts
+++ b/socket/src/index.ts
@@ -1,14 +1,3 @@
-export {
-  Controller,
-  Connection,
-  Connect,
-  Disconnect,
-  GlobalEvent,
-  Event,
-  IO,
-  Socket,
-  Args,
-  Ack
-} from './decorators';
-export { IO_MIDDLEWARE, Middleware } from './middleware';
-export { attachControllers } from './socket';
+export * from './decorators';
+export * from './middleware';
+export * from './socket';

--- a/socket/src/meta.ts
+++ b/socket/src/meta.ts
@@ -1,4 +1,4 @@
-import { Middleware } from './middleware';
+import { Type } from './middleware';
 
 /**
  * Parameter types enum
@@ -38,7 +38,7 @@ export interface Listener {
   methodName: string;
   event: string;
   type: EventType;
-  middleware: Middleware[];
+  middleware: Type[];
 }
 
 /**
@@ -56,7 +56,7 @@ export class SocketMeta {
   /**
    * Middleware
    */
-  middleware: Middleware[] = [];
+  middleware: Type[] = [];
 
   /**
    * Event listeners
@@ -77,9 +77,7 @@ export class SocketMeta {
  * @extends {Object}
  */
 export interface SocketClass extends Object {
-  __socket_meta__: SocketMeta;
-
-  new (...deps: any[]);
+  __socket_meta__?: SocketMeta;
 }
 
 /**

--- a/socket/src/middleware.ts
+++ b/socket/src/middleware.ts
@@ -5,7 +5,25 @@ import { Container, InjectionToken } from '@decorators/di';
  */
 export const IO_MIDDLEWARE = new InjectionToken('IO_MIDDLEWARE');
 
+export interface Type extends Function {
+  new (...args: any[]);
+}
+
 export type NextFunction = (err?: Error) => void;
+
+/**
+ * Server Middleware class interface
+ *
+ * @export
+ * @interface ServerMiddleware
+ */
+export interface ServerMiddleware {
+  use(
+    io: SocketIO.Server | SocketIO.Namespace,
+    socket: SocketIO.Socket,
+    next: NextFunction
+  ): void;
+}
 
 /**
  * Middleware class interface
@@ -14,27 +32,30 @@ export type NextFunction = (err?: Error) => void;
  * @interface Middleware
  */
 export interface Middleware {
-  new (...args: any[]);
-
-  use(...args: any[]): void;
+  use(
+    io: SocketIO.Server | SocketIO.Namespace,
+    socket: SocketIO.Socket,
+    args: any,
+    next: NextFunction
+  ): void;
 }
 
 /**
  * Create request middleware handler that uses class or function provided as middleware
  *
  * @export
- * @param {Middleware} middleware
+ * @param {Type | InjectionToken} middleware
  *
  * @returns {RequestHandler}
  */
-export function middlewareHandler(middleware: Middleware | InjectionToken) {
+export function middlewareHandler(middleware: Type | InjectionToken) {
   return function(...args: any[]): any {
-    let instance: Middleware;
+    let instance: Middleware | ServerMiddleware;
 
     try {
       instance = Container.get(middleware);
     } catch (e) {
-      instance = new (middleware as Middleware)();
+      instance = new (middleware as Type)();
     }
 
     return instance.use.apply(instance, args);
@@ -44,13 +65,12 @@ export function middlewareHandler(middleware: Middleware | InjectionToken) {
 /**
  * Loops through all registered middlewares
  *
- * @description middleware approach
- * @param {MiddlewareFunction[]} fns Array of middleware functions
+ * @param {Type[]} middleware Array of middleware
  * @param {any[]} [args = []] Arguments to pass in
  *
  * @returns {Promise<*>}
  */
-export function executeMiddleware(middleware: Middleware[], args: any[] = []): Promise<any> {
+export function executeMiddleware(middleware: Type[], args: any[] = []): Promise<any> {
   function iteratee(done: (err: Error) => void, i = 0) {
     try {
       middlewareHandler(middleware[i])(...args, (err) => {

--- a/socket/src/socket.ts
+++ b/socket/src/socket.ts
@@ -9,29 +9,29 @@ import {
   SocketClass,
   getMeta
 } from './meta';
-import { executeMiddleware, IO_MIDDLEWARE, middlewareHandler } from './middleware';
+import { executeMiddleware, IO_MIDDLEWARE, middlewareHandler, Type } from './middleware';
 
 /**
  * Attaches controllers to IO server
  *
  * @param {SocketIO.Server} io
- * @param {SocketClass[]} controllers
+ * @param {Type[]} controllers
  */
-export function attachControllers(io: SocketIO.Server, controllers: SocketClass[]) {
+export function attachControllers(io: SocketIO.Server, controllers: Type[]) {
   io.use((socket, next) => middlewareHandler(IO_MIDDLEWARE)(io, socket, next));
 
-  controllers.forEach((controller: SocketClass) => attachController(io, controller));
+  controllers.forEach((controller: Type) => attachController(io, controller));
 }
 
 /**
  * Attach Controller
  *
  * @param {SocketIO.Server} IO
- * @param {SocketClass} Controller
+ * @param {Type} Controller
  */
 function attachController(
   io: SocketIO.Server,
-  controller: SocketClass
+  controller: Type
 ) {
   const instance: SocketClass = Container.get(controller);
   const meta: SocketMeta = getMeta(instance);
@@ -95,13 +95,13 @@ function socketMiddlewareHandler(io: IO, socket: SocketIO.Socket, meta: SocketMe
  *
  * @param {SocketIO.Server|SocketIO.Namespace} io
  * @param {SocketIO.Socket} socket
- * @param {DecoratorsArtifacts} artifacts
+ * @param {Type} controller
  * @param {EventType} type
  */
 function applyEvents(
   io: SocketIO.Server | SocketIO.Namespace,
   socket: SocketIO.Socket,
-  controller: SocketClass,
+  controller: Type,
   type: EventType
 ) {
   const instance: SocketClass = Container.get(controller);

--- a/socket/tsconfig.json
+++ b/socket/tsconfig.json
@@ -15,9 +15,10 @@
     "target": "es6",
     "inlineSources": true,
     "typeRoots": ["node_modules/@types"],
-    "pretty": true
+    "pretty": true,
+    "outDir": "lib"
   },
   "files": [
-    "src/index.ts"
+    "./index.ts"
   ]
 }

